Analytics Hub: Refactor today and yesterday range generation

ThomazFB opened the pull request:
Summary
==========
This is the first of a series of PRs to refactor and improve how the Analytics Hub business logic is implemented. The main goals of this series of changes are to achieve:

- Time zone correctness
- Better error handling and more unit test coverage and safety to the date ranges
- A new UI for the date range selector view
- Previous period comparison support for Custom ranges

Since it will take a bunch of changes to achieve the complete scope, this PR is targeted to a feature branch called `improve/analytics-hub` that will aggregate everything, this way we can review piece by piece while keeping the production version intact.

Also, this PR only addresses the first bit of changes, this new code is not connected to the Hub yet, but if you want to take a peek at the full scope of changes, there's an ongoing draft PR with the entire scope that will be discarded once the feature branch is complete.

You can check it here: https://github.com/woocommerce/woocommerce-android/pull/8039

Changelog
==========
Introduces dedicated class implementation to handle the `Today` and `Yesterday` range generation alongside sample documentation, making it easier to spot how each range business logic works and what's expected. 

Also, adds unit tests checking all of those ranges, including the hours, minutes, and seconds expectations to ensure the Timezone configuration is respected.

How to Test
==========
1. Since the code introduced in this PR is still not connected to the UI, the best way to test this is by tweaking the code and verifying the unit test results.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Define AnalyticsHubTimeRange class

ThomazFB committed:
Define AnalyticsHubDateRangeSelection class

ThomazFB committed:
Define AnalyticsHubRangeSelectionType class

ThomazFB committed:
Define AnalyticsHubTimeRangeData abstract class

ThomazFB committed:
Add range fields to AnalyticsHubTimeRangeData

ThomazFB committed:
Adjust AnalyticsHubTimeRangeData to become an interface

ThomazFB committed:
Introduce basic AnalyticsHubTodayRangeData

ThomazFB committed:
Add initial time range data generation to AnalyticsHubRangeSelectionType

ThomazFB committed:
Merge branch 'trunk' into improve/date-range-strategy

ThomazFB committed:
Initialize AnalyticsHubDateRangeSelection with selection data

ThomazFB committed:
Add auxiliar calendar date operators

ThomazFB committed:
Define AnalyticsHubDateRangeSelectionTest file

ThomazFB committed:
Configure AnalyticsHubDateRangeSelectionTest with initial structure

ThomazFB committed:
Adjust AnalyticsHubDateRangeSelection ranges to be public fields

ThomazFB committed:
Define first test covering today scenario

ThomazFB committed:
Define first test covering yesterday scenario

ThomazFB committed:
Define first test covering last week scenario

ThomazFB committed:
Define first test covering week to date scenario

ThomazFB committed:
Define first test covering last month scenario

ThomazFB committed:
Define first test covering month to date scenario

ThomazFB committed:
Adjust test covering month to date scenario

ThomazFB committed:
Define first test covering last quarter scenario

ThomazFB committed:
Define first test covering quarter to date scenario

ThomazFB committed:
Define first test covering last year scenario

ThomazFB committed:
Define first test covering year to date scenario

ThomazFB committed:
Set calendar parameter to the AnalyticsHubDateRangeSelection

ThomazFB committed:
Implement AnalyticsHubTodayRangeData

ThomazFB committed:
Define AnalyticsHubYesterdayRangeData

ThomazFB committed:
Rename Calendar aux functions

ThomazFB committed:
Enforce calendar correctness to AnalyticsHubDateRangeSelectionTest

ThomazFB committed:
Adjust AnalyticsHubTodayRangeData to not use computed properties for date ranges building

ThomazFB committed:
Rework AnalyticsHubYesterdayRangeData to follow the init pattern

ThomazFB committed:
Modify `generateTimeRangeData` from SelectionType to consider the yesterday case

ThomazFB committed:
Remove all tests besides TODAY and YESTERDAY ones

ThomazFB committed:
Adapt AnalyticsHubDateRangeSelection to use internal SelectionType enum

ThomazFB committed:
Delete AnalyticsHubRangeSelectionType file

ThomazFB committed:
Add documentation to Today and Yesterday range generation

ThomazFB committed:
Fix lint issues

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/DateExt.kt:
```diff
@@ -83,8 +84,22 @@ val Date.pastTimeDeltaFromNowInDays
-fun Date.theDayBeforeIt() =
-    Calendar.getInstance()
-        .apply { time = this@theDayBeforeIt }
+fun Date.theDayBeforeIt(calendar: Calendar = Calendar.getInstance()) =
```
⚙️ Let's specify the return type (I think it's safe to make it non nullable, looking at `Calendar#getTime` implementation)

<img width="1102" alt="image" src="https://user-images.githubusercontent.com/5845095/210340345-99cbd156-b473-4060-9c73-2e5f120356e0.png">

ThomazFB replied:
I agree! I actually fixed that in the upcoming PRs, as you can [see here](https://github.com/woocommerce/woocommerce-android/pull/8039/files#diff-8664e2f7b943994af391a4899dc4a5e6c8e87badb20796dbe371a29f2b67293c). Do you want me to bring those adjustments right away to this PR? The fix will eventually appear in the future.

wzieba replied:
Ok gotcha. No need to add adjustment in this PR 👍

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/DateExt.kt:
```diff
@@ -83,8 +84,22 @@ val Date.pastTimeDeltaFromNowInDays
-fun Date.theDayBeforeIt() =
-    Calendar.getInstance()
-        .apply { time = this@theDayBeforeIt }
+fun Date.theDayBeforeIt(calendar: Calendar = Calendar.getInstance()) =
```
❓ 

1. Could you please clarify why we need an instance of `calendar` to calculate the previous day? It's not very clear to me.
2. Some invocations of `theDayBeforeIt` use a singleton calendar (`Calender#getInstance`) - do you plan to change it in future PRs? If not, how can we determine if we need to pass an instance of a calendar vs. when we should use the singleton instance?

ThomazFB replied:
Hey Wojtek! Since this is the very first PR, let me expand this discussion beyond the context of this single PR and talk about the whole implementation. This way, we can shape all future PRs to follow the outcome of what we decide here, WDYT? I'll provide links to the draft PR with the full implementation to clarify what will be introduced here.

The reason to pass a Calendar parameter instead of using the singleton is to make all the logic around the date ranges generation agnostic of device/local settings and rely solely on a provided calendar configuration. 

This not only makes it easier to define unit tests that check and enforce time zone correctness based on the calendar configuration, but also allows us to define the Date range data classes to work based on the provided date instead of the `today` device configuration.

Doing a dependency inversion of the Calendar reference gives us more testing confidence that every single corner of the range selection generation is following the provided calendar, mitigating risks that we introduce regressions like the ones we did before, and also allow us to unit test specific corner cases regarding calendar date variations through weeks, months and years. [You can verify here how those tests are getting introduced in the full work changelog](https://github.com/woocommerce/woocommerce-android/pull/8039/files#diff-cb5f092046875c3baa01a8cfb5a5ce7586b50f8d411178799f64fceb43968318). 

Also, this allows us to operate all date ranges through the provided calendar reference. Here's a snippet of the custom date range generation from upcoming PR changes that exemplify what I mean:

``` kotlin
class AnalyticsHubCustomRangeData(
    selectedStartDate: Date,
    selectedEndDate: Date,
    calendar: Calendar
) : AnalyticsHubTimeRangeData {
    override val currentRange: AnalyticsHubTimeRange
    override val previousRange: AnalyticsHubTimeRange

    init {
        calendar.time = selectedStartDate
        val currentStart = calendar.startOfCurrentDay()

        calendar.time = selectedEndDate
        val currentEnd = calendar.endOfCurrentDay()

        currentRange = AnalyticsHubTimeRange(
            start = currentStart,
            end = currentEnd
        )

        val dayDifference = selectedEndDate.time - selectedStartDate.time

        calendar.time = selectedStartDate.oneDayAgo(calendar)
        val previousEnd = calendar.endOfCurrentDay()

        calendar.time = Date(previousEnd.time - dayDifference)
        val previousStart = calendar.startOfCurrentDay()

        previousRange = AnalyticsHubTimeRange(
            start = previousStart,
            end = previousEnd
        )
    }
}
``` 

Does this clarify your questions? What's your take on that?

wzieba replied:
I understand the argument with unit tests, I agree it's good to provide the `Calendar` as a dependency!

> The reason to pass a Calendar parameter instead of using the singleton is to make all the logic around the date ranges generation agnostic of device/local settings and rely solely on a provided calendar configuration.

If we plan to provide some constant `java.util.Calendar` configuration in the future (or if we do this already but I missed it), do we plan to remove all instances of `Calendar#getInstance`? I'm concerned that with ambiguous API (you can provide your own `Calendar` or use a default parameter e.g. `theDayBeforeIt`), we won't know what's the correct way.

ThomazFB replied:
Nice! I believe that, at least for now, using the `Calendar.getInstance()` as the production calendar choice is safe since it will retrieve the device Calendar configuration we would want to use for the actual Hub execution. 

The `Calendar` as a dependency is mostly a testability choice. But about the `theDayBeforeIt` function, I believe we can default to the `getInstance()` calendar since this function will adjust the calendar to use the Date receiving the call. This way, we avoid this ambiguity there and leave those Calendar specifics to the Analytics Hub date range operations. WDYT?

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/ranges/AnalyticsHubDateRangeSelection.kt:
```diff
@@ -0,0 +1,39 @@
+class AnalyticsHubDateRangeSelection(
+    selectionType: AnalyticsHubRangeSelectionType,
+    currentDate: Date = Date(),
+    calendar: Calendar = Calendar.getInstance()
```
❓ Here also - why when we need to pass `calendar`, and when can we use the singleton instance?

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/ranges/AnalyticsHubTimeRange.kt:
```diff
@@ -0,0 +1,11 @@
+    val start: Date,
+    val end: Date
+) {
+    val description: String
```
❓ Where this `description` is going to be used? If it's for UI, please consider extracting this to a static helper function instead, as `AnalyticsHubTimeRange` looks like a domain object.

ThomazFB replied:
I should've removed this property, actually. It's not relevant for this PR current context. But the actual implementation is working [like this](https://github.com/woocommerce/woocommerce-android/pull/8039/files#diff-65bcaa40dfb9e13558546b26891499a55533753bb64daa1a3b46709296fb44d7), and tested through [here](https://github.com/woocommerce/woocommerce-android/pull/8039/files#diff-a19d930378596d2616c476dd8ff9c6ddb67c965881c2cd3d0f0f0938be66ec93).

Since we're opening a broader discussion around the full implementation, you can share with me your thoughts on that this way I can adjust the outcome of our discussion for the upcoming PRs.

wzieba replied:
Thanks for the link! Maybe let's discuss this part in the next PR then, sounds good?

ThomazFB replied:
Sounds good, thanks!

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/ranges/data/AnalyticsHubTodayRangeData.kt:
```diff
@@ -0,0 +1,40 @@
+import java.util.Calendar
+import java.util.Date
+
+// Responsible for defining two ranges of data, one starting from the the first second of the current day
```
Minor: 
```suggestion
// Responsible for defining two ranges of data, one starting from the first second of the current day
```

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/ranges/data/AnalyticsHubTodayRangeData.kt:
```diff
@@ -0,0 +1,40 @@
+import java.util.Date
+
+// Responsible for defining two ranges of data, one starting from the the first second of the current day
+// until the same day in the current time and the previous one, starting from the first second of
```
Is that right? I'm not sure what's "(...)day in the current time(...)"

```suggestion
// until the same day in the current timezone and the previous day, starting from the first second of
```

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/ranges/data/AnalyticsHubTodayRangeData.kt:
```diff
@@ -0,0 +1,40 @@
+// yesterday until the same time of that day. E. g.
+//
+// Today: 29 Jul 2022
+// Current range: Jul 29 until Jul 29, 2022
+// Previous range: Jul 28 until Jul 28, 2022
```
As comments says things about time, should we add also time to this example? Is that correct?

```suggestion
// When user requests report at 05:49 PM
// Current range: Jul 29, 00:00 until Jul 29, 05:49 PM 2022
// Previous range: Jul 28, 00:00 until Jul 28, 05:49 PM 2022
```

wzieba reviewed (commented):
Hi @ThomazFB thanks for those improvements. I have a few questions and it was easier to ask them as PR comments than on Slack. Can we discuss things here?

Btw: I think that Analytics Hub was a good opportunity to add Java 8 `java.time` API https://developer.android.com/studio/write/java8-support-table . Maybe there's still a chance to introduce it, WDYT? I think it could benefit us a lot in the future.

ThomazFB committed:
Update WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/ranges/data/AnalyticsHubTodayRangeData.kt

Co-authored-by: Wojciech Zięba <wojtek.zieba@automattic.com>

ThomazFB committed:
Update WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/ranges/data/AnalyticsHubTodayRangeData.kt

Co-authored-by: Wojciech Zięba <wojtek.zieba@automattic.com>

ThomazFB committed:
Update WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/ranges/data/AnalyticsHubTodayRangeData.kt

Co-authored-by: Wojciech Zięba <wojtek.zieba@automattic.com>

ThomazFB committed:
Abandon calendar parameter usage from `theDayBeforeIt` function

ThomazFB committed:
Remove calendar parameter from `theDayBeforeIt` function

ThomazFB committed:
Update AnalyticsHubYesterdayRangeData documentation

ThomazFB commented:
@wzieba, I believe I have addressed all suggestions and discussions from your first round! Thank you so much for diving into a detailed and careful review and starting the discussion around the `java.time` API 🙇 The PR is ready for a second round!

I updated the documentation for both date range data classes, and removed the calendar parameter from the `theDayBeforeIt` function, following our conversations. Let me know what you think.

wzieba reviewed (approved):
Thank you @ThomazFB for the discussion and for addressing improvements, looks good!
